include/util/opening_hours.hpp

```diff
@@ -0,0 +1,645 @@
+#ifndef OPENING_HOURS_HPP
+#define OPENING_HOURS_HPP
+
+//#define BOOST_SPIRIT_DEBUG
+
+#include <boost/date_time/gregorian/gregorian.hpp>
+#include <boost/spirit/include/phoenix.hpp>
+#include <boost/spirit/include/qi.hpp>
+
+#include <boost/io/ios_state.hpp>
+
+#include <algorithm>
+#include <cctype>
+#include <iomanip>
+#include <iterator>
+#include <limits>
+#include <string>
+
+namespace osrm
+{
+namespace util
+{
+
+// Helper classes for "opening hours" format http://wiki.openstreetmap.org/wiki/Key:opening_hours
+// Supported simplified features in CheckOpeningHours:
+// - Year/Month/Day ranges
+// - Weekday ranges
+// - Time ranges
+// Not supported:
+// - Week numbers
+// - Holidays, events, variables dates
+// - Day offsets and periodic ranges
+struct OpeningHours
+{
+    enum Modifier
+    {
+        unknown,
+        open,
+        closed,
+        off,
+        is24_7
+    };
+
+    struct Time
+    {
+        enum Event : unsigned char
+        {
+            invalid,
+            none,
+            dawn,
+            sunrise,
+            sunset,
+            dusk
+        };
+
+        Event event;
+        std::int32_t minutes;
+
+        Time() : event(invalid), minutes(0) {}
+        Time(Event event) : event(event), minutes(0) {}
+        Time(char hour, char min) : event(none), minutes(hour * 60 + min) {}
+        Time(Event event, bool positive, const Time &offset)
+            : event(event), minutes(positive ? offset.minutes : -offset.minutes)
+        {
+        }
+    };
+
+    struct TimeSpan
+    {
+        Time from, to;
+        TimeSpan() = default;
+        TimeSpan(const Time &from_, const Time &to_) : from(from_), to(to_)
+        {
+            if (to.minutes < from.minutes)
+                to.minutes += 24 * 60;
+        }
+
+        bool IsInRange(const struct tm &time, bool &use_curr_day, bool &use_next_day) const
+        {
+            // TODO: events are not handled
+            if (from.event != OpeningHours::Time::none || to.event != OpeningHours::Time::none)
+                return false;
```

Comment:
I guess this is fine for now - tagging conditional attributes with `sunrise` etc. is the mappers fault :+1: